File: LinearAndMixedModel.R
library(dplyr)

list.of.packages <- c("data.table","bit64","optimx","lme4",'languageR','lmerTest','glmnet','lsmeans','car','RDS','DMwR','stringr','nloptr','ggplot2','reshape2','plyr','viridis','gmodels')
if(length(new.packages)) install.packages(new.packages)
lapply(list.of.packages,require,character.only=TRUE)

path="D:\\MCOE_DEV\\MMT_NEW_THINGS\\Ancova\\48\\"

setwd(path)

model_data_final=read.csv("model_data_after_outlier.csv")
names(model_data_final)
model_data_final_1=model_data_final
model_data_final$PERIOD=as.factor(model_data_final$PERIOD)
model_data_final$TEST_CONTROL=as.factor(model_data_final$TEST_CONTROL)

model_data_final$PERIOD=relevel(model_data_final$PERIOD,1)
model_data_final$TEST_CONTROL=relevel(model_data_final$TEST_CONTROL,1)
levels(model_data_final$PERIOD)
######Fixed Effect #########################
formula="DOLACV1 ~ TEST_CONTROL * PERIOD"
#model_full = fit!(lm(formula, train, wts = convert(Array{Float64,1}, train[Symbol("ACV",string(cfg["PID_CNT"]))])))
model_full = lm(formula, model_data_final,weights = model_data_final$ACV1)
summary(model_full)

model_data_final$predicted1=predict(model_full,model_data_final)
model_data_final$residuals=residuals(model_full)
model_data_final_1=model_data_final
model_data_final_1$TEST_CONTROL=ifelse(as.numeric(model_data_final_1$TEST_CONTROL)==2,1,0)
model_data_final_1$PERIOD=ifelse(as.numeric(model_data_final_1$PERIOD)==2,1,0)

model_data_final_1$predicted1 = 24.4117+ (-1.5184* model_data_final_1$TEST_CONTROL)+
  (1.7103* model_data_final_1$PERIOD) + (-0.9224* model_data_final_1$TEST_CONTROL* model_data_final_1$PERIOD ) 
model_data_final_1$residuals=residuals(model_full)
sum(model_data_final_1$predicted1)


model_data_final[model_data_final$STORE==33423,c("STORE","PERIOD","TEST_CONTROL","predicted1","residuals")]
model_data_final_1[model_data_final_1$STORE==33423,c("STORE","PERIOD","TEST_CONTROL","predicted1","residuals")]

model_data_final[model_data_final$STORE==100127,c("STORE","PERIOD","TEST_CONTROL","predicted1","residuals")]
model_data_final_1[model_data_final_1$STORE==100127,c("STORE","PERIOD","TEST_CONTROL","predicted1","residuals")]

cor(model_data_final$predicted1,model_data_final$DOLACV1)
######Store as random effect#############
formula="DOLACV1 ~ TEST_CONTROL * PERIOD + (1|STORE)"
model_full_1 = lmer(formula, model_data_final,weights= model_data_final$ACV1)
summary(model_full_1)
ranef(model_full_1)

model_data_final$predicted1=predict(model_full_1,model_data_final)
a=as.data.frame(ranef(model_full_1))

model_data_final_1=model_data_final
model_data_final_1=merge(model_data_final_1,a,by.x = "STORE",by.y = "grp")

model_data_final_1$TEST_CONTROL=ifelse(as.numeric(model_data_final_1$TEST_CONTROL)==2,1,0)
model_data_final_1$PERIOD=ifelse(as.numeric(model_data_final_1$PERIOD)==2,1,0)

model_data_final_1$predicted1 = 25.1182+ (-1.8359* model_data_final_1$TEST_CONTROL)+
                            (1.6910* model_data_final_1$PERIOD) +
                            (-0.9516* model_data_final_1$TEST_CONTROL* model_data_final_1$PERIOD) +model_data_final_1$condval


model_data_final[model_data_final$STORE==33423,c("STORE","PERIOD","TEST_CONTROL","predicted1")]
model_data_final_1[model_data_final_1$STORE==33423,c("STORE","PERIOD","TEST_CONTROL","predicted1","condval")]

model_data_final[model_data_final$STORE==100127,c("STORE","PERIOD","TEST_CONTROL","predicted1","residuals")]
model_data_final_1[model_data_final_1$STORE==100127,c("STORE","PERIOD","TEST_CONTROL","predicted1","residuals")]

cor(model_data_final$predicted1,model_data_final$DOLACV1)
##############Random Slope of DOLACV0############
formula="DOLACV1 ~ TEST_CONTROL * PERIOD + (0+DOLACV0|STORE)"
model_full_1 = lmer(formula, model_data_final,weights= model_data_final$ACV1)
summary(model_full_1)

ranef(model_full_1)
model_data_final$predicted=predict(model_full_1,model_data_final)

a=as.data.frame(ranef(model_full_1))

model_data_final_1=model_data_final
model_data_final_1=merge(model_data_final_1,a,by.x = "STORE",by.y = "grp")

model_data_final_1$TEST_CONTROL=ifelse(as.numeric(model_data_final_1$TEST_CONTROL)==2,1,0)
model_data_final_1$PERIOD=ifelse(as.numeric(model_data_final_1$PERIOD)==2,1,0)

model_data_final_1=model_data_final_1[model_data_final_1$STORE==33423,]
model_data_final_1$predicted1 = 23.0946+ (-2.1985* model_data_final_1$TEST_CONTROL)+
                            (1.7910* model_data_final_1$PERIOD) +
                            (-0.8331* model_data_final_1$TEST_CONTROL* model_data_final_1$PERIOD) +(model_data_final_1$condval*model_data_final_1$DOLACV0)



model_data_final[model_data_final$STORE==33423,c("STORE","PERIOD","TEST_CONTROL","predicted")]
model_data_final_1[model_data_final_1$STORE==33423,c("STORE","PERIOD","TEST_CONTROL","DOLACV0","predicted1")]

model_data_final[model_data_final$STORE==100127,c("STORE","PERIOD","TEST_CONTROL","predicted")]
model_data_final_1[model_data_final_1$STORE==100127,c("STORE","PERIOD","TEST_CONTROL","predicted1")]

cor(model_data_final$predicted,model_data_final$DOLACV1)
##############Random Intercept with Random Slope of DOLACV0############


formula="DOLACV1 ~ TEST_CONTROL * PERIOD + (DOLACV0|STORE)"
model_full_1 = lmer(formula, model_data_final,weights= model_data_final$ACV1)
summary(model_full_1)

ranef(model_full_1)
model_data_final$predicted=predict(model_full_1,model_data_final)

formula="DOLACV1 ~ TEST_CONTROL * PERIOD + (1+DOLACV0|STORE)"
model_full_2 = lmer(formula, model_data_final,weights= model_data_final$ACV1)
summary(model_full_2)

ranef(model_full_2)
model_data_final$predicted1=predict(model_full_2,model_data_final)
sum(predict(model_full_1,model_data_final))
sum(predict(model_full_2,model_data_final))

cor(predict(model_full_1,model_data_final),model_data_final$DOLACV1)
cor(predict(model_full_2,model_data_final),model_data_final$DOLACV1)

a=as.data.frame(ranef(model_full_2))
b=merge(a[a$term=="(Intercept)",],a[a$term!="(Intercept)",],by.x = "grp", by.y = "grp")
model_data_final_1=model_data_final
model_data_final_1=merge(model_data_final_1,b,by.x = "STORE",by.y = "grp")

model_data_final_1$TEST_CONTROL=ifelse(as.numeric(model_data_final_1$TEST_CONTROL)==2,1,0)
model_data_final_1$PERIOD=ifelse(as.numeric(model_data_final_1$PERIOD)==2,1,0)


head(model_data_final_1)
model_data_final_1$predicted1 = 24.1430+ (-2.1095* model_data_final_1$TEST_CONTROL)+
                            (1.7894* model_data_final_1$PERIOD) +
                            (-0.7986* model_data_final_1$TEST_CONTROL* model_data_final_1$PERIOD)+ model_data_final_1$condval.x +model_data_final_1$condval.y*model_data_final_1$DOLACV0

model_data_final[model_data_final$STORE==33423,c("STORE","PERIOD","TEST_CONTROL","predicted1")]
model_data_final_1[model_data_final_1$STORE==33423,c("STORE","PERIOD","TEST_CONTROL","predicted1")]
sum(model_data_final_1$predicted1)
sum(model_data_final$predicted1)
cor(model_data_final_1$predicted1,model_data_final_1$DOLACV1)

File: LinearAndMixedModel.jl
sort!(model_data_final,cols=[:STORE])
TEST_SUMMARY=DataFrame()
TRAIN_SUMMARY=DataFrame()
df_corr=DataFrame()
a=0
model_data_final[:TEST_CONTROL]=map((x)-> ifelse(x ==2,1,0),model_data_final[:TEST_CONTROL]);
model_data_final[:PERIOD]=map((x)-> ifelse(x ==2,1,0),model_data_final[:PERIOD]);
relevel(model_data_final[:TEST_CONTROL],0)
relevel(model_data_final[:PERIOD],0)
levels(model_data_final[:TEST_CONTROL])
levels(model_data_final[:PERIOD])

for i in 1:100
	
	df_tot_weeks=DataFrame()
	df_tot_weeks=unique(model_data_final[:WEEK])
	test=DataFrame()
	train=DataFrame()
	for str in unique(model_data_final[:STORE])
		#str=50404
		n=floor(length(unique(model_data_final[model_data_final[:STORE].==str,:WEEK]))*0.7)
		df_weeks=DataFrame()
		df_weeks[:WEEK]=sample(model_data_final[model_data_final[:STORE].==str,:WEEK],convert(Int64,n),replace=false)
		train=vcat(train,join(model_data_final[model_data_final[:STORE].==str,:],df_weeks,on=:WEEK))
		test=vcat(test,join(model_data_final[model_data_final[:STORE].==str,:],df_weeks,on=:WEEK,kind=:anti))
	end
    
	s=by(train,[:STORE,:PERIOD],train->length(unique(train[:WEEK])))
	s[:ITR]=i
	TRAIN_SUMMARY=vcat(TRAIN_SUMMARY,s)
	
	s=by(test,[:STORE,:PERIOD],test->length(unique(test[:WEEK])))
	s[:ITR]=i
	TEST_SUMMARY=vcat(TEST_SUMMARY,s)
	
    ########Fixed effect	
	formula=eval(parse(string(string(cfg["MEASURE_ACV"]),string(cfg["PID_CNT"]), "~ TEST_CONTROL * PERIOD ")))
	#model_full = fit!(lm(formula, train, weights = convert(Array{Float64,1}, train[Symbol("ACV",string(cfg["PID_CNT"]))])))
	#model_full = glm(formula, model_data_final,Normal(),IdentityLink(), wts  = convert(Array{Float64,1}, model_data_final[Symbol("ACV",string(cfg["PID_CNT"]))]))
	model_full = glm(formula, train,Normal(),IdentityLink(), wts  = convert(Array{Float64,1}, train[Symbol("ACV",string(cfg["PID_CNT"]))]))
	#model_full = lm(formula, train)
    #model_full = lm(formula, model_data_final)
	FixedEff=DataFrame(transpose(coef(model_full)));names!(FixedEff,[:Intercept,:TEST_CONTROL ,:PERIOD,:TEST_CONTROL_PERIOD])

	predictedDOLACV=StatsBase.predict(model_full,train);
	#predictedDOLACV=StatsBase.predict(model_full,model_data_final[model_data_final[:STORE].==33423,:]);
	a=cor(train[:DOLACV1],predictedDOLACV)
	#cor(model_data_final[:DOLACV1],predictedDOLACV)
	predictedDOLACV=StatsBase.predict(model_full,test);
	a=vcat(a,cor(test[:DOLACV1],predictedDOLACV))

    ########Random effect
	
	formula=eval(parse(string(string(cfg["MEASURE_ACV"]),string(cfg["PID_CNT"]), "~ TEST_CONTROL * PERIOD + (1|STORE)")))
	model_full_1 = fit!(lmm(formula, train, weights = convert(Array{Float64,1}, train[Symbol("ACV",string(cfg["PID_CNT"]))])))
    #model_full_1 = fit!(lmm(formula, model_data_final, weights = convert(Array{Float64,1}, model_data_final[Symbol("ACV",string(cfg["PID_CNT"]))])))
	FixedEff=DataFrame(transpose(coef(model_full_1)));names!(FixedEff,[:Intercept,:TEST_CONTROL ,:PERIOD,:TEST_CONTROL_PERIOD])
	RanEff =DataFrame()
	RanEff[:STORE] =unique(model_full_1.mf.df[:STORE])
	RanEff = hcat(RanEff,DataFrame(transpose(ranef(model_full_1)[1])));names!(RanEff,[:STORE,:RANEFF])

	train1=deepcopy(train[:,[:TEST_CONTROL,:PERIOD,:STORE,:DOLACV1]])
	#train1=deepcopy(model_data_final[:,[:TEST_CONTROL,:PERIOD,:STORE,:DOLACV1]])
	#train1[:TEST_CONTROL]=map((x)-> ifelse(x ==2,1,0),train1[:TEST_CONTROL]);
	#train1[:PERIOD]=map((x)-> ifelse(x ==2,1,0),train1[:PERIOD]);
	train1=join(train1,FixedEff, kind = :cross)
	train1=join(train1,RanEff,on = :STORE, kind = :inner)
	train1[:predicted] = train1[:Intercept].+ (train1[:TEST_CONTROL_1].* train1[:TEST_CONTROL]).+ (train1[:PERIOD_1].* train1[:PERIOD]) .+ (train1[:TEST_CONTROL_PERIOD].* train1[:TEST_CONTROL].* train1[:PERIOD]) .+ train1[:RANEFF]
	a=vcat(a,cor(train1[:DOLACV1],train1[:predicted]))
	writetable("RanEff_ranef.csv",RanEff,header=true);
	#train1[train1[:STORE].==33423,:]


	test1=deepcopy(test[:,[:TEST_CONTROL,:PERIOD,:STORE,:DOLACV1]])
	#test1[test1[:test1_CONTROL].==2,:TEST_CONTROL]=0
	#test1[:PERIOD]=map((x)-> ifelse(x ==2,1,0),test1[:PERIOD]);
	test1=join(test1,FixedEff, kind = :cross)
	test1=join(test1,RanEff,on = :STORE, kind = :inner)
	test1[:predicted] = test1[:Intercept].+ (test1[:TEST_CONTROL_1].* test1[:TEST_CONTROL]).+ (test1[:PERIOD_1].* test1[:PERIOD]) .+ (test1[:TEST_CONTROL_PERIOD].* test1[:TEST_CONTROL].* test1[:PERIOD]) .+ test1[:RANEFF]
	a=vcat(a,cor(test1[:DOLACV1],test1[:predicted]))
	
	df_corr=vcat(df_corr,DataFrame(transpose(a)))

	########Random slope
	formula=eval(parse(string(string(cfg["MEASURE_ACV"]),string(cfg["PID_CNT"]), "~ TEST_CONTROL * PERIOD + (0+",cfg["MEASURE_ACV"],"0|STORE)")))
	model_full_1 = fit!(lmm(formula, train, weights = convert(Array{Float64,1}, train[Symbol("ACV",string(cfg["PID_CNT"]))])))
	#model_full_1 = fit!(lmm(formula, model_data_final, weights = convert(Array{Float64,1}, model_data_final[Symbol("ACV",string(cfg["PID_CNT"]))])))

	FixedEff=DataFrame(transpose(coef(model_full_1)));names!(FixedEff,[:Intercept,:TEST_CONTROL ,:PERIOD,:TEST_CONTROL_PERIOD])
	RanEff =DataFrame()
	RanEff[:STORE] =unique(model_full_1.mf.df[:STORE])
	RanEff = hcat(RanEff,DataFrame(transpose(ranef(model_full_1)[1])));names!(RanEff,[:STORE,:SLOPE])

	train1=deepcopy(train[:,[:TEST_CONTROL,:PERIOD,:STORE,:DOLACV1,:DOLACV0]])
	#train1=model_data_final[:,[:TEST_CONTROL,:PERIOD,:STORE,:DOLACV1,:DOLACV0]]
	#train1[:TEST_CONTROL]=map((x)-> ifelse(x ==2,1,0),train1[:TEST_CONTROL]);
	#train1[:PERIOD]=map((x)-> ifelse(x ==2,1,0),train1[:PERIOD]);
	train1=join(train1,FixedEff, kind = :cross)
	train1=join(train1,RanEff,on = :STORE, kind = :inner)
	train1[:predicted] = train1[:Intercept].+ (train1[:TEST_CONTROL_1].* train1[:TEST_CONTROL]).+ (train1[:PERIOD_1].* train1[:PERIOD]) .+ (train1[:TEST_CONTROL_PERIOD].* train1[:TEST_CONTROL].* train1[:PERIOD]) .+ train1[:SLOPE] .* train1[:DOLACV0]
	a=vcat(a,cor(train1[:DOLACV1],train1[:predicted]))
	writetable("RanEff_slope.csv",RanEff,header=true);
	#train1[train1[:STORE].==33423,:]
    

	test1=deepcopy(test[:,[:TEST_CONTROL,:PERIOD,:STORE,:DOLACV1,:DOLACV0]])
	#test[:TEST_CONTROL]=map((x)-> ifelse(x ==2,1,0),test[:TEST_CONTROL]);
	#test[:PERIOD]=map((x)-> ifelse(x ==2,1,0),test[:PERIOD]);
	test1=join(test1,FixedEff, kind = :cross)
	test1=join(test1,RanEff,on = :STORE, kind = :inner)
	test1[:predicted] = test1[:Intercept].+ (test1[:TEST_CONTROL_1].* test1[:TEST_CONTROL]).+ (test1[:PERIOD_1].* test1[:PERIOD]) .+ (test1[:TEST_CONTROL_PERIOD].* test1[:TEST_CONTROL].* test1[:PERIOD]) .+ test1[:SLOPE] .* test1[:DOLACV0]
	a=vcat(a,cor(test1[:DOLACV1],test1[:predicted]))
	
	df_corr=vcat(DataFrame(transpose(a)))
	
	########Random Intercept and Random slope
	formula=eval(parse(string(string(cfg["MEASURE_ACV"]),string(cfg["PID_CNT"]), "~ TEST_CONTROL * PERIOD + (1+",cfg["MEASURE_ACV"],"0|STORE)")))
	model_full_1 = fit!(lmm(formula, train, weights = convert(Array{Float64,1}, train[Symbol("ACV",string(cfg["PID_CNT"]))])))
	#model_full_1 = fit!(lmm(formula, model_data_final, weights = convert(Array{Float64,1}, model_data_final[Symbol("ACV",string(cfg["PID_CNT"]))])))

	FixedEff=DataFrame(transpose(coef(model_full_1)));names!(FixedEff,[:Intercept,:TEST_CONTROL ,:PERIOD,:TEST_CONTROL_PERIOD])
	RanEff =DataFrame()
	RanEff[:STORE] =unique(model_full_1.mf.df[:STORE])
	RanEff = hcat(RanEff,DataFrame(transpose(ranef(model_full_1)[1])));names!(RanEff,[:STORE,:RANEFF,:SLOPE])

	train1=deepcopy(train[:,[:TEST_CONTROL,:PERIOD,:STORE,:DOLACV1,:DOLACV0]])
	#train1=model_data_final[:,[:TEST_CONTROL,:PERIOD,:STORE,:DOLACV1,:DOLACV0]]
	train1[:TEST_CONTROL]=map((x)-> ifelse(x ==2,1,0),train1[:TEST_CONTROL]);
	train1[:PERIOD]=map((x)-> ifelse(x ==2,1,0),train1[:PERIOD]);
	train1=join(train1,FixedEff, kind = :cross)
	train1=join(train1,RanEff,on = :STORE, kind = :inner)
	train1[:predicted] = train1[:Intercept].+ (train1[:TEST_CONTROL_1].* train1[:TEST_CONTROL]).+ (train1[:PERIOD_1].* train1[:PERIOD]) .+ (train1[:TEST_CONTROL_PERIOD].* train1[:TEST_CONTROL].* train1[:PERIOD]) .+ train1[:RANEFF] .+ train1[:SLOPE].* train1[:DOLACV0]
	a=vcat(a,cor(train1[:DOLACV1],train1[:predicted]))
	writetable("RanEff_int_slope.csv",RanEff,header=true);
	#train1[train1[:STORE].==33423,:]


	test1=deepcopy(test[:,[:TEST_CONTROL,:PERIOD,:STORE,:DOLACV1,:DOLACV0]])
	#test1[test1[:TEST_CONTROL].==2,:TEST_CONTROL]=0
	#test1[:PERIOD]=map((x)-> ifelse(x ==2,1,0),test1[:PERIOD]);
	test1=join(test1,FixedEff, kind = :cross)
	test1=join(test1,RanEff,on = :STORE, kind = :inner)
	test1[:predicted] = test1[:Intercept].+ (test1[:TEST_CONTROL_1].* test1[:TEST_CONTROL]).+ (test1[:PERIOD_1].* test1[:PERIOD]) .+ (test1[:TEST_CONTROL_PERIOD].* test1[:TEST_CONTROL].* test1[:PERIOD]) .+ test1[:RANEFF]
	a=vcat(a,cor(test1[:DOLACV1],test1[:predicted]))
	
	df_corr=vcat(DataFrame(transpose(a)))
	
end


writetable(root*"/TRAIN_SUMMARY.csv",TRAIN_SUMMARY,header = true);
writetable(root*"/TEST_SUMMARY.csv",TEST_SUMMARY,header = true);
writetable(root*"/df_corr.csv",df_corr,header = true);
